server/neptune/workflows/internal/terraform/runners/job_runner.go

```diff
@@ -0,0 +1,98 @@
+package runners
+
+import (
+	"strings"
+
+	"github.com/hashicorp/go-version"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/github"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/steps"
+	"go.temporal.io/sdk/workflow"
+)
+
+// CustomRunner runs custom run steps.
+type StepRunner interface {
+	Run(
+		ctx workflow.Context,
+		step steps.Step,
+		repo github.Repo,
+		commit github.Commit,
+		tfVersion *version.Version,
+		projectName string,
+		repoRelDir string,
+		path string,
+		envs map[string]string,
+	) (string, error)
+}
+
+// JobRunner runs a deploy job
+type JobRunner interface {
+	Run(
+		ctx workflow.Context,
+		job steps.Job,
+		repo github.Repo,
+		commit github.Commit,
+		tfVersion *version.Version,
+		projectName string,
+		repoRelDir string,
+		path string,
+	) (string, error)
+}
+
+func NewJobRunner(
+	runStepRunner StepRunner,
+	envStepRunner StepRunner,
+) JobRunner {
+	jobRunner := &jobRunner{}
+	jobRunner.RunRunner = runStepRunner
+	jobRunner.EnvRunner = envStepRunner
+
+	return jobRunner
+}
+
+func (r *jobRunner) Run(
+	ctx workflow.Context,
+	job steps.Job,
+	repo github.Repo,
+	commit github.Commit,
+	tfVersion *version.Version,
+	projectName string,
+	repoRelDir string,
+	path string,
```

Comment:
I don't want to make the same mistake as atlantis, we shouldn't be passing in like 50 things to functions. Structure data in a way that encapsulates and provides only necessary access patterns.

In this case you can imagine having something like 
```
type RootInstance struct {
    Name string
    Path string
    Repo *RepoInstance
}

func (r *Root) RelativePathFromRepo() {
    filepath.RelativePath(Path, r.Repo.Path)
}

type RepoInstance struct {
   HeadCommit github.Commit
   Repo github.Repo
   Path string
}

type ExecutionContext struct {
    Path string
    Envs map[string]string

    // allows you to forgo passing in context.Context separately and also allows us to just use this object in place of everywhere we need context.Context.
    Context
}
```

Reply:
Thanks for the suggestion. This makes much more sense :) I have tried modeling it similar to your suggestion. Let me know what you think